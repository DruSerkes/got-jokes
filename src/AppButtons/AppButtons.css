.ShowFavoritesButton {
  position: relative;
}

.ShowFavoritesButton::after {
  align-items: center;
  background-color: whitesmoke;
  border-radius: 50%;
  color: #0F0F0F;
  content: attr(data-favorites);
  display: flex;
  font-size: 0.75rem;
  justify-content: center;
  height: 1.25rem;
  position: absolute;
  right: -0.75rem;
  top: -0.75rem;
  width: 1.25rem;
  z-index: 10;
}