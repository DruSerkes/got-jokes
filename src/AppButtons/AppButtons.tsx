import * as React from 'react';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import Button from 'react-bootstrap/Button';
import { SaveJoke } from '../SaveJoke/SaveJoke';
import { RemoveFavorites } from '../RemoveFavorites/RemoveFavorites';
import { jokeData } from '../types';
import { TweetButton } from '../TweetButton/TweetButton';

interface AppButtonsProps {
  viewingFavorites: boolean;
  isLoading: boolean;
  joke: jokeData | undefined;
  favorites: jokeData[];
  setJoke: React.Dispatch<React.SetStateAction<jokeData | undefined>>;
  setViewingFavorites: React.Dispatch<React.SetStateAction<boolean>>;
  setFavorites: React.Dispatch<jokeData[]>;
  clearFavorites: () => void;
}

export const AppButtons = ({ viewingFavorites, isLoading, joke, favorites, setJoke, setViewingFavorites, setFavorites, clearFavorites }: AppButtonsProps) => {
  const handleShowFavorites = () => setViewingFavorites(!viewingFavorites);
  const handleGetJoke = () => {
    setJoke(undefined);
    setViewingFavorites(false);
  };
  const saveFavoriteJoke = () => {
    if (!joke) return;
    const newFavorites: jokeData[] = [joke, ...favorites];
    setFavorites(newFavorites);
    setJoke(undefined)
  };
  return (
    <ButtonGroup aria-label="Main App Buttons" className="Buttons">
      <Button
        variant="primary"
        onClick={handleGetJoke}
        disabled={isLoading ? true : false}
        className='my-4'>
        Make Me Laugh!
      </Button>

      <SaveJoke joke={joke} viewingFavorites={viewingFavorites} saveFavoriteJoke={saveFavoriteJoke} />

      <Button
        variant="secondary"
        onClick={handleShowFavorites}
        className='my-4'>
        {!viewingFavorites ? 'Show my Favorites' : 'Hide my Favorites'}
      </Button>

      {joke && !viewingFavorites && <TweetButton joke={joke} bootstrapClasses={['my-4']} />}

      <RemoveFavorites clearFavorites={clearFavorites} haveFavorites={!!favorites.length} viewingFavorites={viewingFavorites} />
    </ButtonGroup>
  );
};