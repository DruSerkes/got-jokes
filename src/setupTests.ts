// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';

export const testJoke = {
    "error": false,
    "category": 'dark',
    "type": 'single',
    "joke": 'test joke',
    "flags": {
        "nsfw": false,
        "religious": false,
        "political": false,
        "racist": false,
        "sexist": false,
        "explicit": false
    },
    "id": 13,
    "safe": true,
    "lang": 'en',
    'setup': 'test setup',
    'delivery': 'test delivery'
};

export const setJoke = jest.fn(() => console.log('setting joke'));
export const setViewingFavorites = jest.fn(() => console.log('viewing favorites'));
export const setFavorites = jest.fn(() => console.log('setting favorites'));
export const clearFavorites = jest.fn(() => console.log('clearing favorites'));
export const saveFavoriteJoke = jest.fn(() => console.log('saving favorite joke'));