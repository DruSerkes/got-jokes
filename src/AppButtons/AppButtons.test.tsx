import * as React from 'react';
import { render } from '@testing-library/react';
import { AppButtons } from './AppButtons';
import { testJoke, setJoke, setViewingFavorites, setFavorites, clearFavorites } from '../setupTests';

describe('AppButtons tests', () => {

    it('should render and match snapshot', () => {
        const { asFragment } = render(<AppButtons viewingFavorites={false} isLoading={false} joke={testJoke} favorites={[]} setJoke={setJoke} setViewingFavorites={setViewingFavorites} setFavorites={setFavorites} clearFavorites={clearFavorites} />);
        expect(asFragment()).toMatchSnapshot();
    });

}); 