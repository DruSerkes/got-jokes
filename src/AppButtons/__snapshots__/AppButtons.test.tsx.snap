// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppButtons tests should render and match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="Main App Buttons"
    class="Buttons btn-group"
    role="group"
  >
    <button
      class="my-4 btn btn-primary"
      type="button"
    >
      Make Me Laugh!
    </button>
    <button
      class="my-4 btn btn-info"
      type="button"
    >
      Save to Favorites
    </button>
    <button
      class="my-4 btn btn-secondary"
      type="button"
    >
      Show my Favorites
    </button>
    <a
      class="my-4 TweetButton btn btn-outline-dark"
      href="https://twitter.com/intent/tweet?text=test joke"
      rel="noopener noreferrer"
      target="_blank"
      title="tweet this joke"
    />
  </div>
</DocumentFragment>
`;
